[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40038 [quant][graphmode] Support another use pattern of mean
* #39919 [quant] Support general op modules with inplace options
* #39925 [quant][graphmode] Support quantizing `repeat`
* #39924 [quant][graphmode] Support squeeze/unsqueeze
* #39923 [quant][graphmode] Run RemoveRedundantDequantize in the end
* #39915 [quant][graphmode] Pass debug option into insert_quant_dequant pass
* #39911 [quant][graphmode] Use quantized::batch_norm in graph mode
* #39910 [quant] add quantized::batch_norm
* **#39895 [quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

jerryzh168 committed:
[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/quantization/helper.cpp:
```diff
@@ -295,9 +295,10 @@ std::vector<Value*> getPassThroughInputs(Value* v) {
     }
     return inputs;
-  } else if (n->kind() == prim::ListUnpack) {
+  } else if (n->kind() == prim::ListUnpack || n->kind() == prim::TupleUnpack) {
```
Can you explain why you return input(0) here?

jerryzh168 replied:
input(0) is the input of the list unpack/ tuple unpack. which is a list/tuple. this function is to propagate observed property from input to output, in this case we can propagate the observed property from the list/tuple to their output, which are the unpacked values.

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]

vkuzo reviewed (approved):
lg

jerryzh168 committed:
Update on "[quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22009854](https://our.internmc.facebook.com/intern/diff/D22009854)

[ghstack-poisoned]
